### PR TITLE
enable per-channel location weight scaling

### LINKED ISSUE #2292: Allow for per channel location weights in loss computation
### Describe the task. It can be a feature, documentation, etc.

Currently there is only a global (per stream) option to apply location weighting (e.g. cosine_latitude).
This is to explore a smooth transition between uniform weighting and custom weighting that can be specified per channel.

Idea:
channel_location_weight = (1-p)-p*latitude_weight

for p=1 full weight, for p=0 uniform

Background: for some (especially low pressure level) variables most dynamics are around the poles

### Hedgedoc URL, if you are keeping notes, plots, logs in hedgedoc.

_No response_

### Area

- [ ] datasets, data readers, data preparation and transfer
- [x] model
- [x] science
- [ ] infrastructure and engineering
- [ ] evaluation, export and visualization
- [ ] documentation
- [ ] performance

### PR DESCRIPTION
**Pull request by jpolz**
## Description

Backward compatible per channel scaling of the location loss weighting.

Example config:
```yaml
location_weight: cosine_latitude
location_weight_fraction:
  q: 0.0    # no lat weighting
  u: 0.5    # half lat weighting
  t: 1.0    # full (same as omitting)
  # unlisted channels default to 1.0
channel_weights:
  q: 2.0    # applied after, as before
```

## Issue Number

Closes https://github.com/ecmwf/WeatherGenerator/issues/2292

Is this PR a draft? Mark it as draft.

## Checklist before asking for review

-   [ ] I have performed a self-review of my code
-   [ ] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [ ] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

### CODE REVIEW

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 1]
```diff
@@ -114,7 +114,12 @@ def kernel_crps(
 
     # apply point weighting
     if weights_points is not None:
-        kcrps_locs_chs = kcrps_locs_chs * weights_points
+        if weights_points.dim() == 1:
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py:117]: Can't we handle this without and if by broadcasting the shape in the uniform case?

File: src/weathergen/train/loss_modules/loss_functions.py [hunk 2]
```diff
@@ -195,7 +200,12 @@ def lp_loss(
         torch.abs(torch.where(mask_nan, target, 0) - torch.where(mask_nan, pred, 0)), p_norm
     )
     if weights_points is not None:
-        diff_p = (diff_p.transpose(1, 0) * weights_points).transpose(1, 0)
+        if weights_points.dim() == 1:
```
Comment (clessig) [inline src/weathergen/train/loss_modules/loss_functions.py:203]: Same as above

**Review by clessig — COMMENTED** [review]
For me the concept of a fractional location weight is strange since it corrects for something that is either there or not. I see why one might want to control the weighting but conceptually it is still questionable.

In terms of the config, the pattern matching is problematic e.g. for `z`. So having an explicit list with all channels explicitly and the explicit weights is much preferable.
